tensorflow/python/ops/array_ops.py

```diff
@@ -2721,11 +2721,13 @@ def batch_gather(params, indices, name=None):
       raise ValueError("batch_gather does not allow indices with unknown "
                        "shape.")
     batch_indices = indices
+    dtype = indices.dtype
     accum_dim_value = 1
```

Comment:
Initialize it to dtype(1) as well?

Reply:
I don't understand what you said. You can see the referenced issue for the purpose of this change.

Reply:
Well, this function do support dynamic shapes, because `params_shape = shape(params)` is a Tensor and `params_shape[dim]` is a slice and all these are dynamic. I experimented dynamic shapes, and it does work. Also, I found a Dimension * Tensor returns a Tensor, and, therefore, your proposal has no effect after first multiplication of `accum_dim_value`. By the way, all these slicing, multiplication and `gen_math_ops._range` inserts nodes in the graph. I believe using `gen_math_ops.cast` is not a bad idea.
If we really use `Dimension` all the way, `gen_math_ops._range` will anyway convert `Dimension` to `Tensor`. Also, since `batch_indices` is a Tensor, we must convert all these to `Tensor` in the end.